pageserver/src/tenant/blob_io.rs

```diff
@@ -72,14 +73,22 @@ impl<'a> BlockCursor<'a> {
                 len_buf.copy_from_slice(&buf[off..off + 4]);
                 off += 4;
             }
-            len_buf[0] &= !LEN_COMPRESSION_BIT_MASK;
+            let bit_mask = if self.read_compressed {
+                !LEN_COMPRESSION_BIT_MASK
+            } else {
+                0x7f
+            };
+            len_buf[0] &= bit_mask;
             u32::from_be_bytes(len_buf) as usize
         };
         let compression_bits = first_len_byte & LEN_COMPRESSION_BIT_MASK;
 
         let mut tmp_buf = Vec::new();
         let buf_to_write;
-        let compression = if compression_bits <= BYTE_UNCOMPRESSED {
+        let compression = if compression_bits <= BYTE_UNCOMPRESSED || !self.read_compressed {
+            if compression_bits > BYTE_UNCOMPRESSED {
+                warn!("reading key above future limit ({len} bytes)");
+            }
```

Comment:
Doesn't this mean that we are reading a previously compressed as uncompressed?

Shouldn't we also or alternatively check if the read bytes start with the [zstd fourcc/magic](https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#zstandard-frames)?

Reply:
Anyway, I am having a hard time parsing this compression_bits. I get it that here it can be the u32 after anding it with `0x0f` or `0x7f` which means -- aha, it is never masked.

Related note: On line L332R341 the LEN_COMPRESSION_BIT_MASK is used as literal `0xf0`:

```
assert_eq!(len_buf[0] & 0xf0, 0);
```

Ok... So perhaps I know understand. Possible compression_bits are:

```
match compression_bits >> 4 {
  0 => /* image layer written before the compression support */,
  1..8 => /* undefined or written before compression support and too large, which we warn here? */,
  8 => /* uncompressed blob */,
  9 => /* zstd */,
  10..=15 => { /* reserved */ }
}
```

If this is correctly understood, then okay maybe ... The `compression_bits > BYTE_COMPRESSED` just looks so off, in my mind a bitfield doesn't support ordered comparison. It'd be nice to have enums and matches for these.

Reply:
> Doesn't this mean that we are reading a previously compressed as uncompressed?

If the compression setting is disabled, yes. This has the consequence that we can't turn off compression easily any more, but I think it's okay to have it for a while, after which point we'll (mostly) revert this PR.

>  It'd be nice to have enums and matches for these

`match` is not good for that as `FOOBAR =>` is equivalent to a variable capture.

Reply:
Is my match correct? Did you test if this is hit? Will that be used as a success criteria for the compression support? If so, what is the plan to read all image layers?


Other questions remain, magic/fourcc instead of reserving more bits?

Reply:
> Is my match correct? 

almost correct, the `1..8` range doesn't have the highest bit set so it's an indicator for small uncompressed values.

```rust
match compression_bits >> 4 {
  0..8 => /* small, uncompressed value below 128 bytes */,
  8 => /* uncompressed blob */,
  9 => /* zstd */,
  10..=15 => /* reserved or written before compression support and too large, which we warn here */
  _ => unreachable!("u4"),
}
```

> Will that be used as a success criteria for the compression support?

What do you mean, can you expand?

>  If so, what is the plan to read all image layers?

??

> magic/fourcc instead of reserving more bits?

What do you mean by that? I do not think we should autodetect zstd magic here.

Reply:
Christian wants this to be merged as-is, doing that now.

Reply:
> success criteria
> plan to read all image layers

Christian mentioned these in https://github.com/neondatabase/neon/pull/8238#issuecomment-2206472870. I was wondering similarly what are the next steps.

> > magic/fourcc
>
> What do you mean by that? I do not think we should autodetect zstd magic here.

zstd always starts the compressed bytes with the same 4 magic bytes. I was thinking if we should instead use that knowledge instead of awkwardly reserving bits, as I had no idea what was the plan for the next step.. But yeah, seems there is a plan after all.

Reply:
For future reference, the slack thread where I proposed to merge asap is https://neondb.slack.com/archives/C074JQU7NER/p1720021276202079